[Model Element] Implement lazy loading and unloading based on viewport intersection

amarchenko opened the pull request:
#### 3391caa9afee4eadf7547dc93143b531eac64e45
<pre>
[Model Element] Implement lazy loading and unloading based on viewport intersection
<a href="https://bugs.webkit.org/show_bug.cgi?id=295640">https://bugs.webkit.org/show_bug.cgi?id=295640</a>
149973010

Reviewed by Ada Chan.

This change implements lazy loading for model elements to improve performance by
only loading models when they are visible in the viewport and unloading them
when they scroll out of view. That helps to reduce memory usage and improve page
load performance for pages with many model elements.

Since WebCore/Sources.txt is changed, missing headers added here and there to fix the build.

Key changes:
- Add LazyLoadModelObserver class that uses IntersectionObserver to track when
  model elements enter/exit the viewport
- Models now only begin loading when they intersect the viewport, not immediately
  when sources are set
- Models are unloaded when they leave the viewport to free memory
- Update visibility logic to consider both page visibility and viewport intersection
- Rename reloadModelTimer to loadModelTimer to better reflect its purpose

* LayoutTests/model-element/model-element-animations-replace-sources.html:
* LayoutTests/model-element/model-element-bounding-box.html:
* LayoutTests/model-element/model-element-contents-layer-updates-with-clipping.html:
* LayoutTests/model-element/model-element-contents-layer-updates.html:
* LayoutTests/model-element/model-element-entity-transform.html:
* LayoutTests/model-element/model-element-environment-map-ready.html:
* LayoutTests/model-element/model-element-lazy-loading-unloading-expected.txt: Added.
* LayoutTests/model-element/model-element-lazy-loading-unloading.html: Added.
* LayoutTests/model-element/model-element-on-hidden-page-with-delays-expected.txt: Added.
* LayoutTests/model-element/model-element-on-hidden-page-with-delays.html: Added.
* LayoutTests/model-element/model-element-ready.html:
* LayoutTests/model-element/model-element-stagemode.html:
* LayoutTests/model-element/resources/model-element-test-utils.js:
(async waitForModelState):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::resume):
(WebCore::HTMLModelElement::visibilityStateChanged):
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::startLoadModelTimer):
(WebCore::HTMLModelElement::loadModelTimerFired):
(WebCore::HTMLModelElement::didFailLoading):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::didUnload):
(WebCore::HTMLModelElement::shouldDeferLoading const):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::stop):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
(WebCore::HTMLModelElement::sourceRequestResource):
(WebCore::HTMLModelElement::viewportIntersectionChanged):
(WebCore::HTMLModelElement::isModelDeferred const):
(WebCore::HTMLModelElement::isModelLoading const):
(WebCore::HTMLModelElement::isModelLoaded const):
(WebCore::HTMLModelElement::isModelUnloading const):
(WebCore::HTMLModelElement::isModelUnloaded const):
(WebCore::HTMLModelElement::modelElementStateForTesting const):
(WebCore::HTMLModelElement::startReloadModelTimer): Deleted.
(WebCore::HTMLModelElement::reloadModelTimerFired): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/LazyLoadModelObserver.cpp: Added.
(WebCore::LazyLoadModelObserver::observe):
(WebCore::LazyLoadModelObserver::unobserve):
(WebCore::LazyLoadModelObserver::intersectionObserver):
* Source/WebCore/Modules/model-element/LazyLoadModelObserver.h: Added.
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
* Source/WebCore/Modules/permissions/NavigatorPermissions.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::scriptExecutionContext const):
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.cpp:
* Source/WebCore/Modules/speech/LocalDOMWindowSpeechSynthesis.cpp:
* Source/WebCore/Modules/webaudio/ConvolverNode.cpp:
* Source/WebCore/Modules/webaudio/IIRFilterNode.cpp:
* Source/WebCore/Modules/webaudio/OfflineAudioDestinationNode.cpp:
* Source/WebCore/Modules/webaudio/WaveShaperNode.cpp:
* Source/WebCore/Modules/webdatabase/SQLResultSetRowList.cpp:
* Source/WebCore/Modules/webdriver/NavigatorWebDriver.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::lazyLoadModelObserver):
* Source/WebCore/dom/Document.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::modelElementState):
(WebCore::Internals::isModelElementIntersectingViewport):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::unloadModelTimerFired):

Canonical link: <a href="https://commits.webkit.org/297264@main">https://commits.webkit.org/297264@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c2162aa2a65743e624d5ad0719d17f7f28b8e1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30744 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117109 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31425 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39326 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84448 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114025 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25107 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100006 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64894 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110509 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24451 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18151 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60929 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94486 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18216 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119999 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38127 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28334 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93389 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38503 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96283 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93213 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23756 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38290 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16031 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34108 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38016 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43492 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37680 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39383 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c95309ce5070b1a2cefa1b89054773747f94b4a3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47771

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b6e9b16f83cf2f7f19cff62bafc9154d0834c01. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47771

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -135,7 +141,7 @@ void HTMLModelElement::suspend(ReasonForSuspension reasonForSuspension)
 {
     RELEASE_LOG(ModelElement, "%p - HTMLModelElement::resume()", this);
-    startReloadModelTimer();
+    startloadModelTimer();
```
Nit: should be `startLoadModelTimer()`.

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -180,19 +186,23 @@ void HTMLModelElement::visibilityStateChanged()
 {
     setSourceURL(selectModelSource());
+
+    // Deferred: m_model == nullptr && m_resource == nullptr
```
Maybe the comment can benefit with more elaboration.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -466,26 +467,32 @@ void HTMLModelElement::reloadModelPlayer()
-    RELEASE_LOG(ModelElement, "%p - HTMLModelElement: Reloading model player after becoming visible", this);
+    RELEASE_LOG(ModelElement, "%p - HTMLModelElement: Timer fired for loading model", this);
+
+    if (m_model == nullptr && m_resource == nullptr) {
```
Ditto about calling `isModelDeferred` here.

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
 #endif
 
+void HTMLModelElement::sourceRequestResource()
+{
```
Should we check or add an assertion that `m_sourceURL` is not empty?

amarchenko replied:
Added source URL check.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+bool HTMLModelElement::isModelDeferred() const
+{
+    return m_model == nullptr && m_resource == nullptr;
```
Ditto about using `!` instead of comparing against `nullptr`

achan00 replied:
Also, I'd like to understand better the expectation of this method when `m_sourceURL` is empty. Right now it looks like it'd always return `true` when there's no source URL. Is that expected?

amarchenko commented:
Also addressed errors from mac-safer-cpp.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+bool HTMLModelElement::isModelLoading() const
+{
+    return m_isIntersectingViewport && ((m_model == nullptr && m_resource != nullptr) || (m_model && !m_modelPlayer) || (m_modelPlayer && m_modelPlayer->isPlaceholder()));
```
Should we call `isVisible()` rather than checking `m_isIntersectingViewport` here?

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+bool HTMLModelElement::isModelLoaded() const
+{
+    return m_isIntersectingViewport && m_modelPlayer && !m_modelPlayer->isPlaceholder();
```
Ditto

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+bool HTMLModelElement::isModelUnloading() const
+{
+    return !m_isIntersectingViewport && m_modelPlayer && !m_modelPlayer->isPlaceholder();
```
Ditto (`!isVisible()`)

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+bool HTMLModelElement::isModelUnloaded() const
+{
+    return !m_isIntersectingViewport && (m_model && (!m_modelPlayer || m_modelPlayer->isPlaceholder()));
```
Ditto.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -400,6 +400,9 @@ static ResourceError toResourceError(String payload, Model& model)
         return;
 
+    if (m_loader)
+        m_loader->cancel();
```
Can set `m_loader` to `nullptr` also.

amarchenko replied:
Done.

amarchenko commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1419,6 +1445,78 @@ size_t HTMLModelElement::externalMemoryCost() const
+
+String HTMLModelElement::modelElementStateForTesting() const
+{
+    if (document().page()->shouldDisableModelLoadDelaysForTesting())
```
My thinking: when timer's are disabled transitions between states happen extremely quick, it's hard to write a reliable layout test (much higher chance of flakiness). Keeping in mind that main purpose of this method is testing, I decided to add that check to prevent timing nightmare for a user. I may add a comment about that.

amarchenko replied:
Agree, removing the check.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd0d7b51f7f2928c502b306a5893d8da4e4ffb19. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47771

achan00 commented on LayoutTests/model-element/model-element-lazy-loading-unloading.html:
```diff
@@ -0,0 +1,94 @@
+
+    window.scrollTo(0, 0)
+
+    await waitForModelState(model, "Unloaded");
```
I think we should call `internals.disableModelLoadDelaysForTesting();` for these layout tests. Waiting for multiple seconds to make a successful check is a bit long for layout tests.

amarchenko replied:
I halved the timeout for this method, would like to keep the code tested with timers to keep it as close as possible to the real use-cases.

achan00 replied:
Halving the timeout doesn't change the time needed to run the test successfully, just that it'd time out faster for the failure case.

I think it'd be cleaner to have a separate `internals` method to control whether model loads should be deferred until they are scrolled into view, and repurpose `internals.disableModelLoadDelaysForTesting()` such that it can take in the delay to use for the timer. Then you can specify a shorter timer delay for the tests to run faster. You can file a bug to address that later.

achan00 commented on LayoutTests/model-element/model-element-on-hidden-page-with-delays.html:
```diff
@@ -0,0 +1,64 @@
+    const intermediateState = internals.modelElementState(model);
+    assert_true(intermediateState !== "Loaded", "Model should not be loaded");
+
+    await sleepForSeconds(4.5);
```
Is there any way we can shorten these sleeps? Layout tests are expected to run as quickly as it can with short waits (which is why `internals.disableModelLoadDelaysForTesting` was added)

amarchenko replied:
Reduced.

achan00 commented on Source/WebCore/Modules/model-element/LazyLoadModelObserver.cpp:
```diff
@@ -0,0 +1,110 @@
+/*
```
Should this file be included in `WebCore/Sources.txt`?

amarchenko replied:
Apparently yes. Added.

amarchenko commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -180,19 +186,22 @@ void HTMLModelElement::visibilityStateChanged()
 {
     setSourceURL(selectModelSource());
+
+    if (isModelDeferred() && isConnected() /* && !m_sourceURL.isEmpty() */)
```
Per our offline discussion, we actually don't need to subscribe for intersection observation here. The element can rely on `insertedIntoAncestor() / removedFromAncestor()` methods. Removed.

amarchenko commented:
After changing Source.txt the order files in unified groups has changed as well. That led to a number of `use of incomplete type XYZ` in different files. To fix that, corresponding headers have been added.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c551c87dcc39be2675231eb30e5b1f545d157eb1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47771

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9c2162aa2a65743e624d5ad0719d17f7f28b8e1b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47771

amarchenko committed:
[Model Element] Implement lazy loading and unloading based on viewport intersection
https://bugs.webkit.org/show_bug.cgi?id=295640
149973010

Reviewed by Ada Chan.

This change implements lazy loading for model elements to improve performance by
only loading models when they are visible in the viewport and unloading them
when they scroll out of view. That helps to reduce memory usage and improve page
load performance for pages with many model elements.

Since WebCore/Sources.txt is changed, missing headers added here and there to fix the build.

Key changes:
- Add LazyLoadModelObserver class that uses IntersectionObserver to track when
  model elements enter/exit the viewport
- Models now only begin loading when they intersect the viewport, not immediately
  when sources are set
- Models are unloaded when they leave the viewport to free memory
- Update visibility logic to consider both page visibility and viewport intersection
- Rename reloadModelTimer to loadModelTimer to better reflect its purpose

* LayoutTests/model-element/model-element-animations-replace-sources.html:
* LayoutTests/model-element/model-element-bounding-box.html:
* LayoutTests/model-element/model-element-contents-layer-updates-with-clipping.html:
* LayoutTests/model-element/model-element-contents-layer-updates.html:
* LayoutTests/model-element/model-element-entity-transform.html:
* LayoutTests/model-element/model-element-environment-map-ready.html:
* LayoutTests/model-element/model-element-lazy-loading-unloading-expected.txt: Added.
* LayoutTests/model-element/model-element-lazy-loading-unloading.html: Added.
* LayoutTests/model-element/model-element-on-hidden-page-with-delays-expected.txt: Added.
* LayoutTests/model-element/model-element-on-hidden-page-with-delays.html: Added.
* LayoutTests/model-element/model-element-ready.html:
* LayoutTests/model-element/model-element-stagemode.html:
* LayoutTests/model-element/resources/model-element-test-utils.js:
(async waitForModelState):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::resume):
(WebCore::HTMLModelElement::visibilityStateChanged):
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::startLoadModelTimer):
(WebCore::HTMLModelElement::loadModelTimerFired):
(WebCore::HTMLModelElement::didFailLoading):
(WebCore::HTMLModelElement::isVisible const):
(WebCore::HTMLModelElement::didUnload):
(WebCore::HTMLModelElement::shouldDeferLoading const):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::stop):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
(WebCore::HTMLModelElement::sourceRequestResource):
(WebCore::HTMLModelElement::viewportIntersectionChanged):
(WebCore::HTMLModelElement::isModelDeferred const):
(WebCore::HTMLModelElement::isModelLoading const):
(WebCore::HTMLModelElement::isModelLoaded const):
(WebCore::HTMLModelElement::isModelUnloading const):
(WebCore::HTMLModelElement::isModelUnloaded const):
(WebCore::HTMLModelElement::modelElementStateForTesting const):
(WebCore::HTMLModelElement::startReloadModelTimer): Deleted.
(WebCore::HTMLModelElement::reloadModelTimerFired): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/LazyLoadModelObserver.cpp: Added.
(WebCore::LazyLoadModelObserver::observe):
(WebCore::LazyLoadModelObserver::unobserve):
(WebCore::LazyLoadModelObserver::intersectionObserver):
* Source/WebCore/Modules/model-element/LazyLoadModelObserver.h: Added.
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
* Source/WebCore/Modules/permissions/NavigatorPermissions.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::scriptExecutionContext const):
* Source/WebCore/Modules/screen-wake-lock/WakeLockSentinel.cpp:
* Source/WebCore/Modules/speech/LocalDOMWindowSpeechSynthesis.cpp:
* Source/WebCore/Modules/webaudio/ConvolverNode.cpp:
* Source/WebCore/Modules/webaudio/IIRFilterNode.cpp:
* Source/WebCore/Modules/webaudio/OfflineAudioDestinationNode.cpp:
* Source/WebCore/Modules/webaudio/WaveShaperNode.cpp:
* Source/WebCore/Modules/webdatabase/SQLResultSetRowList.cpp:
* Source/WebCore/Modules/webdriver/NavigatorWebDriver.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::lazyLoadModelObserver):
* Source/WebCore/dom/Document.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::modelElementState):
(WebCore::Internals::isModelElementIntersectingViewport):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::unloadModelTimerFired):

Canonical link: https://commits.webkit.org/297264@main
